### PR TITLE
ENH: optimize: dense redundancy removal routine optimizations

### PR DESCRIPTION
**Pull request by mdhaber**
This PR includes four efficiency improvements for `linprog`'s dense redundancy removal routine. For easier review, please consider each commit separately.

This PR also improves the redundancy removal test suite: previously only the SVD-based RR method was being tested; now all three RR routines go through the same tests.  Unfortunately, GH is not good at detecting the changes. _ALL_ I did to `test__remove_redundancy.py` is:

- add all the separate tests to a class, `RRCommonTests`
- Create three subclasses, `TestRRSVD`. `TestRRPivotDense`, and `TestRRPivotSparse`, each with their own definition of method `rr`
- Replace calls to `_remove_redundancy` with `self.rr`

The fastest way to verify that no unintended changes were introduced is probably to perform the changes in reverse to get back the original file.

@hakeemo I had these in mind for a while. Thanks for the nudge to revisit the redundancy removal routines! Some of these ideas might be helpful if you make the change you suggested in gh-11215. Or, if you'd like, I can just add those changes in this PR.

@Kai-Striega I think made a similar mistake as in gh-10602! How did you get rid of the unintended changes to the submodule?

### CODE REVIEW

**Comment by mdhaber** [comment]
I used `%timeit` to measure the execution time of `_remove_redundancy_dense` on Netlib benchmark problems.

On BORE3D:
With master: `102 ms ± 4.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`
This PR: `35.4 ms ± 2.57 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)`

On SHELL:
With master: 2 s ± 150 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
This PR: `628 ms ± 41.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

On CYCLE:
With master: `21 s ± 2.4 s per loop (mean ± std. dev. of 7 runs, 1 loop each)`
This PR: `5.93 s ± 550 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

So the speedup is ~3x across wide range of problem sizes.

**Comment by mdhaber** [comment]
TravisCI 20603.7 seems to be the only real failure. Apparently `np.eye` did not accept an `order` argument in the past.
TravisCI 20603.8 failed because there seems to be trouble getting SuiteSparse.
Azure Windows builds failed to install MinGW.

**Review by Kai-Striega — COMMENTED** [review]
Changes look good to me 👍. I've left a few comments for you to take a look at.  I haven't tried to duplicate the performance benchmarks on my machine. 

W.R.T. the submodule I would recommend using [git revert](https://www.atlassian.com/git/tutorials/undoing-changes/git-revert) to take out the changes. In the past I would have recommended trying to rebase your changes, but I've changed my opinion on that because it changes the history.

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Originally, the algorithm would hit an exception in the `try` block because `lu` was `None`, then `lu` would get created in the `except` by `lu = scipy.linalg.lu_factor(B)`. But `B` is initially the identity matrix, so it is easy (and faster) to define it manually.

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Good idea.

File: scipy/optimize/tests/test__remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/tests/test__remove_redundancy.py]: There hasn't been a problem for the past two years. Leave it alone?

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [inline scipy/optimize/_remove_redundancy.py]: I guess I still want to use `eye` here. @Kai-Striega would you prefer I do `zeros`+`fill_diagonal`?

Comment (peterbell10) [reply scipy/optimize/_remove_redundancy.py]: You can also use `np.eye(m).T` to get fortran order.

**Comment by josef-pkt** [comment]
I just saw this and it might be useful for statsmodels (but I'm not working on anything related right now)

Is it possible to expose some of this as a public function? 
AFAICS it is all private in optimize.

**Comment by mdhaber** [comment]
@josef-pkt I've thought about that. Yes, it's currently all private, but it could be made public. It would probably belong in `scipy.linalg`? There are a handful of related PRs I'd like to finish up, including improvements to the sparse redundancy removal routine. Maybe after that I'll write up an issue suggesting a public interface.

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Thanks. Would it behave identically, though?
I went ahead and replaced with `np.zeros` + `np.fill_diagonal`, as it's just as fast and a little more explicit about the Fortran order.

Comment (peterbell10) [reply scipy/optimize/_remove_redundancy.py]: >Would it behave identically, though?

The identity matrix is symmetric so it is still an identity after transposing. Further, `.T` just creates a strided view. You can try for yourself:
```python
>>> x = np.eye(4).T
>>> x
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])
>>> x.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```

Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: I know, and I saw the flags. The only one that was different was 'OWNDATA', which was expected, as I knew it would be a view. Just confirming, as I don't know everything under the hood. Thanks.

File: scipy/optimize/tests/test__remove_redundancy.py
Comment (Kai-Striega) [reply scipy/optimize/tests/test__remove_redundancy.py]: I'm happy just to leave it.

File: scipy/optimize/_remove_redundancy.py
Comment (Kai-Striega) [reply scipy/optimize/_remove_redundancy.py]: >  @Kai-Striega would you prefer I do zeros+fill_diagonal?

I'm happy with either. zeros+fill_diagonal was just a suggestion as the (then) existing code fails the test suite.

Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Haha yeah sounds like a good reason to change it.

Comment (Kai-Striega) [reply scipy/optimize/_remove_redundancy.py]: either way, `np.eye(m, order='F')` is still the most legible option (in my opinion). It would probably be worth adding a comment why it was *not* used and to update it when the NumPy version is increased.

Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Oops, yeah, I meant to do that.

**Comment by mdhaber** [comment]
I'd like to wait for gh-10762 to merge before this one so we can see the change in benchmarks.

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: I should have just waited : ) The min NumPy is now 1.14.5.

**Comment by mdhaber** [comment]
Mostly good news:
```
       before           after         ratio
     [cb42f2f6]       [86251d86]
     <master>         <test_rr_improvments>
-             0.0            -46.0      n/a  optimize_linprog.Netlib_RR.track_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'CYCLE')
-         209±3ms        137±0.7ms     0.65  optimize_linprog.Netlib.time_netlib(('interior-point', {'sparse': False}), 'SCORPION')
-     1.10±0.06ms         706±30μs     0.64  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'AFIRO')
-      6.50±0.4ms       3.36±0.2ms     0.52  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'BLEND')
-     11.3±0.08ms      5.57±0.02ms     0.49  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'RECIPE')
-     9.22±0.06ms      4.47±0.03ms     0.48  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'VTP-BASE')
-      15.4±0.5ms      7.40±0.06ms     0.48  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'FINNIS')
-      7.44±0.01s       3.53±0.06s     0.47  optimize_linprog.Netlib.time_netlib(('interior-point', {'sparse': False}), 'QAP8')
-       56.0±20ms       25.8±0.1ms     0.46  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'BORE3D')
-         123±1ms       55.5±0.2ms     0.45  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'SCORPION')
-         157±2ms       68.6±0.4ms     0.44  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'ETAMACRO')
-      1.56±0.07s          592±3ms     0.38  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'SIERRA')
-      1.33±0.02s          470±6ms     0.35  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'SHELL')
-        61.3±3ms      21.2±0.08ms     0.35  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'SCSD6')
-         176±1ms       58.0±0.5ms     0.33  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'WOOD1P')
-        189±70ms       57.7±0.3ms     0.31  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'DEGEN2')
-      3.22±0.09s          897±4ms     0.28  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x10c9aa200>, 'PILOTNOV')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
But after the changes it's calculating a different rank for `CYCLE`. I need to investigate before this can be merged.

**Comment by mdhaber** [comment]
Replacing `dtrsv` with `dtrsm` makes this a little slower, but it fixes the rank for `CYCLE`.
```
       before           after         ratio
     [91c2b4c9]       [413da825]
     <stash~1>        <dense_rr_speedup>
+      1.33±0.02s       1.51±0.01s     1.13  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x119833950>, 'DEGEN3')
+      2.04±0.02s       2.25±0.04s     1.11  optimize_linprog.Netlib_RR.time_netlib_rr(<function _remove_redundancy_dense at 0x119833950>, 'QAP8')
-          -129.0           -129.0     1.00  optimize_linprog.Netlib_RR.track_netlib_rr(<function _remove_redundancy_sparse at 0x1198339e0>, 'WOOD1P')
+           -46.0              0.0    -0.00  optimize_linprog.Netlib_RR.track_netlib_rr(<function _remove_redundancy_dense at 0x119833950>, 'CYCLE')
```

**Comment by mdhaber** [comment]
@josef-pkt Would you be willing to review this? This and gh-11249 would help me get to gh-11414.

**Comment by mdhaber** [comment]
@Kai-Striega If you have a moment...

File: scipy/optimize/_remove_redundancy.py
Comment (Kai-Striega) [inline scipy/optimize/_remove_redundancy.py]: This isn't blocking, but, could we try to avoid generic exceptions?

File: scipy/optimize/tests/test__remove_redundancy.py
Comment (Kai-Striega) [inline scipy/optimize/tests/test__remove_redundancy.py]: A quick question, there are now three routines ``_remove_redundancy``, ``_remove_redundancy_dense`` and ``_remove_redundancy_sparse``, obviously I would use the dense routine for dense arrays and the sparse routine for sparse arrays. But under what circumstances would I use the general ``_remove_redundancy`` over the dense or sparse versions?

**Review by Kai-Striega — APPROVED** [review]
@mdhaber sorry about the delay, this kinda fell under the radar with the grant work / my actual work. Changes look good to me, and the tests are passing.

At first glance the tests seem to have had some major reworks - but that looks like it's only changing them to a class so that they can be re-run with the different parameters more easily, correct?

File: scipy/optimize/tests/test__remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/tests/test__remove_redundancy.py]: They are just poorly named. `_remove_redundancy` is another dense routine that can be useful if the dimension of the nullspace is small. If you can merge this, gh-11249 would be a perfect place to fix the names.

Comment (Kai-Striega) [reply scipy/optimize/tests/test__remove_redundancy.py]: Ok I'll merge it now.

File: scipy/optimize/_remove_redundancy.py
Comment (mdhaber) [reply scipy/optimize/_remove_redundancy.py]: Yes, I strive for that in general. If I remember correctly: here, as in `_get_solver` of [`_linprog_ip.py`](https://github.com/scipy/scipy/blob/master/scipy/optimize/_linprog_ip.py), there are a variety of exceptions that can be raised when factoring, and I wasn't sure whether I had them all. I should revisit this, so I've added it to gh-9671.

**Pull request merged**
